File: examples/checks/calidor.py
# Independent Python checks for the calidor example.
from .common import run_fragment_checks

CHECKS = [
    ('the trusted precomputed HMAC signature verifies', ['Usage is permitted only for purpose "heatwave_response" and the envelope expires at 2026-07-18T21:00:00+00:00.']),
    ('payload hash matches the source envelope digest', ['reason.txt : The gateway keeps raw indoor heat, vulnerability, and prepaid-energy data local, derives a priority-support signal, and shares only a scoped heatwave-response envelope']),
    ('the shared insight strips raw heat, vulnerability, credit, and meter-trace terms', ['The gateway desensitizes local heat, vulnerability, and prepaid-energy stress into an expiring municipal support insight.']),
    ('the insight has a device scope, event scope, and expiry', ['question : Is the Calidor heat-response system allowed to use a narrow household support insight for heatwave response, and if so which support package should it recommend?']),
    ('heatwave-response use is authorized before the insight expires', ['expires at : 2026-07-18T21:00:00+00:00']),
    ('the heat alert is active', ['unsafe indoor heat : active - 31.4°C for 9 hours reaches the 30.0°C/6 hour threshold']),
    ('indoor heat is unsafe for long enough to trigger support', ['Tenant-screening reuse is blocked by a prohibition on odrl:distribute for purpose "tenant_screening".']),
    ('the active-need count reaches the priority threshold', ['energy constraint : active - €3.2 prepaid credit is at or below the €5.0 limit']),
    ('the recommended package is within budget and covers every required capability', ['pkg:VOUCHER  : reject, cost=€12, covers 2/4 required capabilities; within budget']),
    ('the lowest-cost eligible package is chosen', ['Selected package "Calidor Priority Cooling Bundle" covers bill_credit, cooling_kit, transport, welfare_check.']),
    ('the deletion duty is scheduled before expiry', ['vulnerability present : active - the local gateway sees heat-sensitive and mobility flags']),
    ('reuse for tenant screening is prohibited', ['Envelope HMAC-SHA-256: e635c7c1991742a5c36992fc0da32a7abc80b32aa5777a1142adaab55183681c']),
    ('all source-level policy, scope, signature, and package checks pass', ['pkg:DELUXE   : reject, cost=€28, covers all required capabilities; over budget by €8']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/calidor.go
package main

import (
	"fmt"
	"strings"
)

func checkCalidor(ctx *Context) []Check {
	d := ctx.M()
	needs := calidorActiveNeeds(d)
	active := 0
	for _, v := range needs {
		if v {
			active++
		}
	}
	required := []string{"bill_credit", "cooling_kit", "transport", "welfare_check"}
	selected := chooseCapabilityPackage(maps(d["Packages"]), num(d["MaxPackageCostEur"]), required)
	reported := parseCalidorAnswer(ctx.Answer)
	insight := asMap(d["Insight"])
	policy := asMap(d["Policy"])
	sig := asMap(d["Signature"])
	allowed := str(d["RequestAction"]) == str(policy["PermissionAction"]) &&
		str(d["RequestPurpose"]) == str(policy["PermissionPurpose"]) &&
		str(policy["PermissionTarget"]) == str(insight["ID"]) &&
		parseTime(str(d["CityAuthAt"])).Before(parseTime(str(d["GatewayExpiresAt"]))) &&
		active >= integer(d["MinimumActiveNeedCount"]) &&
		selected != nil
	deleteBefore := parseTime(str(d["CityDutyAt"])).Before(parseTime(str(d["GatewayExpiresAt"])))
	serialized := strings.ToLower(str(insight["SerializedLowercase"]))
	localTerms := []string{"heat_sensitive_condition", "mobility_limitation", "prepaid"}
	omitted := true
	for _, t := range localTerms {
		omitted = omitted && !contains(serialized, t)
	}
	cheaperReject := true
	deluxeOverBudget := false
	for _, pkg := range maps(d["Packages"]) {
		if selected != nil && integer(pkg["CostEur"]) < integer(selected["CostEur"]) {
			cheaperReject = cheaperReject && !setContainsAll(setOf(sarr(pkg["Capabilities"])), required)
		}
		if str(pkg["PackageID"]) == "pkg:DELUXE" {
			deluxeOverBudget = integer(pkg["CostEur"]) > integer(d["MaxPackageCostEur"])
		}
	}
	scopeOK := true
	for _, k := range []string{"ScopeDevice", "ScopeEvent", "Municipality", "CreatedAt", "ExpiresAt"} {
		scopeOK = scopeOK && str(insight[k]) != ""
	}
	selectedOK := selected != nil && str(selected["Name"]) == "Calidor Priority Cooling Bundle"
	return []Check{
		{"four active heat-response needs are recomputed from the local signals", active == 4 && reported["active"] == "4"},
		{"the insight threshold of three active needs is met", active >= integer(d["MinimumActiveNeedCount"]) && integer(d["MinimumActiveNeedCount"]) == integer(insight["ThresholdCount"])},
		{"the lowest-cost eligible package covering all required capabilities is selected", selectedOK && reported["pkg"] == str(selected["Name"])},
		{"the selected package fits inside the €20 budget", selected != nil && reported["cost"] == fmt.Sprint(integer(selected["CostEur"])) && integer(selected["CostEur"]) == 18 && integer(selected["CostEur"]) <= integer(d["MaxPackageCostEur"])},
		{"cheaper packages are rejected because they do not cover all capabilities", cheaperReject},
		{"the deluxe package is rejected because it is over budget", deluxeOverBudget},
		{"policy permission authorizes heatwave-response use before expiry", allowed && contains(ctx.Answer, "decision : ALLOWED")},
		{"tenant-screening reuse is prohibited by the policy", str(policy["ProhibitionAction"]) == "odrl:distribute" && str(policy["ProhibitionPurpose"]) == "tenant_screening"},
		{"deletion duty is scheduled before envelope expiry", deleteBefore},
		{"vulnerability flags and local raw stress signals are omitted from the serialized insight", omitted},
		{"reported signature metadata matches the trusted precomputed input", reported["hash"] == str(sig["PayloadHashSHA256"]) && reported["hmac"] == str(sig["SignatureHMAC"]) && str(sig["Algorithm"]) == "HMAC-SHA256"},
		{"scope metadata is explicit for device, event, municipality, creation, and expiry", scopeOK},
	}
}
